pkg/notify/notify.go

```diff
@@ -3,166 +3,218 @@ package notify
+	Search(jql string, options *jira.SearchOptions) ([]jira.Issue, *jira.Response, error)
+	GetTransitions(id string) ([]jira.Transition, *jira.Response, error)
+
+	Create(issue *jira.Issue) (*jira.Issue, *jira.Response, error)
+	UpdateWithOptions(issue *jira.Issue, opts *jira.UpdateQueryOptions) (*jira.Issue, *jira.Response, error)
+	DoTransition(ticketID, transitionID string) (*jira.Response, error)
+}
+
+// Receiver wraps a specific Alertmanager receiver, with its configuration and templates to notify given ticket service.
 type Receiver struct {
-	conf   *config.ReceiverConfig
-	tmpl   *template.Template
-	client *jira.Client
+	logger log.Logger
+	client jiraIssueService
+	// TODO(bwplotka): Consider splitting receiver config with ticket service details.
+	conf *config.ReceiverConfig
+	tmpl *template.Template
+
+	timeNow func() time.Time
 }
 
-// NewReceiver creates a Receiver using the provided configuration and template.
-func NewReceiver(c *config.ReceiverConfig, t *template.Template) (*Receiver, error) {
-	tp := jira.BasicAuthTransport{
-		Username: c.User,
-		Password: string(c.Password),
-	}
-	client, err := jira.NewClient(tp.Client(), c.APIURL)
+// NewReceiverWithClient creates a Receiver using the provided configuration, template and ticketClientService.
+func NewReceiverWithClient(logger log.Logger, c *config.ReceiverConfig, t *template.Template, client jiraIssueService) *Receiver {
+	return &Receiver{logger: logger, conf: c, tmpl: t, client: client, timeNow: time.Now}
+}
+
+// Notify translates alert to a ticket.
+func (r *Receiver) Notify(data *alertmanager.Data) (bool, error) {
+	project, err := r.tmpl.Execute(r.conf.Project, data)
 	if err != nil {
-		return nil, err
+		return false, errors.Wrap(err, "render project")
 	}
 
-	return &Receiver{conf: c, tmpl: t, client: client}, nil
-}
+	issueGroupLabel := toGroupTicketLabel(data.GroupLabels)
 
-// Notify implements the Notifier interface.
-func (r *Receiver) Notify(data *alertmanager.Data, logger log.Logger) (bool, error) {
-	project := r.tmpl.Execute(r.conf.Project, data, logger)
-	if err := r.tmpl.Err(); err != nil {
-		return false, err
+	issue, retry, err := r.issueToReuse(project, issueGroupLabel)
+	if err != nil {
+		return retry, err
 	}
-	// Looks like an ALERT metric name, with spaces removed.
-	issueLabel := toIssueLabel(data.GroupLabels)
 
-	issue, retry, err := r.search(project, issueLabel, logger)
+	// We want up to date title no matter what.
+	// This allows reflecting current group state if desired by user e.g {{ len $.Alerts.Firing() }}
+	issueSummary, err := r.tmpl.Execute(r.conf.Summary, data)
 	if err != nil {
-		return retry, err
+		return false, errors.Wrap(err, "render summary")
 	}
 
 	if issue != nil {
+		// Update summary if needed.
+		if issue.Fields.Summary != issueSummary {
+			retry, err := r.updateSummary(issue.Key, issueSummary)
+			if err != nil {
+				return retry, err
+			}
+		}
+
+		if len(data.Alerts.Firing()) == 0 {
```

Comment:
I believe this check should go before the summary update. Else, with a misconfigured receiver that also sends "resolved" notifications, all of your Jira issues' summaries will end up looking like "TasksMissing: RESOLVED backend prod". (Or whatever your summary template says, but likely either "resolved" or "0 alerts firing".)

I see now that the check is actually duplicated below. Moving it before the `if issue != nil` block, would also get rid of that.

Reply:
Well it's intentially like this. Misconfigured or not you want to see if it's still firing or not, how to achieve that? And it's not minor use case. It's quite important IMO (: Let's discuss this more. Maybe some configuration on this would nice to tell what you want or not? 

Any way we can catch misconfiguration? Do we have to? What misconfiguration do you mean exactly?

Reply:
Furthermore, the code is more complicated and the results more confusing due to the "zero alerts firing" checks. There are 3 directions this may go in:
1. JIRA issue exists and is open: summary is updated to "0 alerts firing", issue is left open (even though there's nothing to do, according to the summary).
2. JIRA issue exists and is resolved: issue is left as resolved, but summary is updated after the issue was closed.
3. No matching JIRA issue exists: no issue is created, nothing is updated.

All in all, not very intuitive or consistent. Why is it that we're updating the summary of a closed issue, but if the issue doesn't exist or was closed long enough ago, we don't do anything?